proto/kv-connect.md

```diff
@@ -0,0 +1,448 @@
+MUST first narrow down the list of endpoints to only include endpoints that have
+a `consistency` property that is equal to `strong`. From this list, the client
+MAY make requests to any endpoint, but the client SHOULD prefer endpoints that
+have lower latency.
+
+The client sends an _Atomic Write Request_ to the server by sending a POST
+request to the `<endpointUrl>/atomic_write` endpoint, where `<endpointUrl>` is
+the resolved URL of the selected endpoint. The request body MUST be a Protobuf
+message in the format `com.deno.kv.datapath.AtomicWriteRequest` (defined in the
+`schema/datapath.proto` file in this repository).
+
+Upon receiving the POST request, the server MUST verify that the request body
+adheres to the Protobuf schema for the `com.deno.kv.datapath.AtomicWriteRequest`
+message. If the request body is invalid, the server MUST respond with a 4xx
+class HTTP status response and a plain text body containing a human readable
+error message.
+
+The server MAY perform quota checks on the request. If the request is rejected
+due to quota limits, the server MUST respond with a 4xx class HTTP status
+response and a plain text body containing a human readable error message.
+
+The server MUST perform the requested atomic write operation. The the server
+MUST respond with a 200 OK response. The response MUST include a `Content-Type`
+header with the value `application/x-protobuf`. The response body MUST be a
+Protobuf message in the format `com.deno.kv.datapath.AtomicWriteResponse`. If
+the server is unable to perform the write because the database is not available
+from this server, the `status` field MUST be set to `AW_WRITE_DISABLED`. If a
+write operation fails due to a check conflict, the `status` field MUST be set to
+`AW_CHECK_FAILED`. If the write operation succeeds, the `status` field MUST be
+set to `AW_SUCCESS`. If the request succeeds, the server MUST include the
+`versiontamp` field with the versionstamp of the write operation.
+
+If the client fails to receive a response from the server due to a network
+error, or a 5xx class HTTP status, the client SHOULD retry the request using an
+exponential backoff strategy with unlimited retries.
+
+The client MUST verify that the response is a 200 OK response. If the response
+is a 4xx class error response, the client MUST fatally abort the request and
+display the error message to the user. If the response is a 200 OK response, the
+client MUST verify that the response body adheres to the Protobuf schema for the
+`com.deno.kv.datapath.AtomicWriteResponse` message. If the response body is
+invalid, the client MUST fatally abort the request and display the error message
+to the user.
+
+If the response has a `status` field set to `AW_WRITE_DISABLED`, the client
+SHOULD perform a metadata exchange with the server to get a new list of
+endpoints, and then retry the request.
+
+If the response has a `status` field set to `AW_CHECK_FAILED`, the client MUST
```

Comment:
I believe we also now return the conflict key for `AW_CHECK_FAILED`, which should also be mentioned here